ntoskrnl/ke/amd64/xstate.c

```diff
@@ -0,0 +1,246 @@
+
+        /* Check if compaction is enabled */
+        if (XStateConfig->CompactionEnabled)
+        {
+            /* Align the offset, if needed */
+            if (ExtStateComponent.Ecx.Bits.Aligned)
+            {
+                NextUserOffset = ALIGN_UP(NextUserOffset, 64);
+                XStateConfig->AlignedFeatures |= ComponentBit;
+            }
+
+            /* Save and update the offset */
+            XStateConfig->Features[Component].Offset = NextUserOffset;
+            NextUserOffset += ExtStateComponent.Size;
+        }
+        else
+        {
+            /* Not compacted, we use the offset specified by the CPUID */
+            XStateConfig->Features[Component].Offset = ExtStateComponent.Offset;
+
+            /* Update the offset, using the highest value */
+            CurrentOffset = XStateConfig->Features[Component].Offset +
+                            XStateConfig->Features[Component].Size;
+            NextUserOffset = max(NextUserOffset, CurrentOffset);
+        }
+    }
+
+     /* Save the sizes */
+    XStateConfig->Size = NextUserOffset;
+    XStateConfig->AllFeatureSize = NextSupervisorOffset;
+
+    /* Save the features to be enabled */
+    XStateConfig->EnabledFeatures = SupportedUserMask;
+    XStateConfig->EnabledSupervisorFeatures = SupportedSupervisorMask;
+
+
+    /* Query CPUID again to get the full size based on enabled components */
+    __cpuidex(ExtStateMain.AsInt32,
+              CPUID_EXTENDED_STATE,
+              CPUID_EXTENDED_STATE_MAIN_LEAF);
+
+    /* Get the full size */
+    XStateConfig->AllFeatureSize = ExtStateMain.Ecx;
+}
+
+/*!
+ *   \brief Validates the provided extended state configuration against the global one
+ *
+ *   \param XStateConfig - Pointer to a XSTATE_CONFIGURATION structure containing the configuration
+ */
+static
+VOID
+ValidateXStateConfig(
+    _In_ PXSTATE_CONFIGURATION XState)
+{
+    PXSTATE_CONFIGURATION GlobalXState = &SharedUserData->XState;
+
+    if ((XState->EnabledFeatures != GlobalXState->EnabledFeatures) ||
+        (XState->EnabledSupervisorFeatures != GlobalXState->EnabledSupervisorFeatures) ||
+        (XState->Size != GlobalXState->Size) ||
+        (XState->AllFeatureSize != GlobalXState->AllFeatureSize))
+    {
+        /* Invalid features */
+        KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+    }
+
+    for (ULONG i = 0; i < MAXIMUM_XSTATE_FEATURES; i++)
+    {
+        if ((XState->Features[i].Size != GlobalXState->Features[i].Size) ||
+            (XState->Features[i].Offset != GlobalXState->Features[i].Offset) ||
+            (XState->AllFeatures[i] != GlobalXState->AllFeatures[i]))
+        {
+            /* Invalid features */
+            KeBugCheck(MULTIPROCESSOR_CONFIGURATION_NOT_SUPPORTED);
+        }
+    }
+}
+
+/*!
+ *   \brief Initializes the extended state configuration for the current processor
+ *
+ *   \param ProcessorNumber - Number of the current processor
+ */
+VOID
```

Comment:
You might want to place this function into the kernel's INIT section so it can be discarded after `KiSwitchToBootStack()`.
```suggestion
CODE_SEG("INIT")
VOID
```

Reply:
You can remove that, when you implement support for CPU hot-plugging 😄